### PR TITLE
config/config.go

### LINKED ISSUE #1454: Respect XDG_CONFIG_HOME when deciding where to store config file
It'd be great if delve respected `$XDG_CONFIG_HOME` when defined instead of storing the config in the user's home directory.

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables

1. What version of Delve are you using (`dlv version`)?
1.1.0

2. What version of Go are you using? (`go version`)?
go version go1.11 linux/amd64

3. What operating system and processor architecture are you using?
Ubuntu amd64

4. What did you do?
Run any `delve` command

5. What did you expect to see?
Configuration file stored at `/home/greg/.config/dlv`

6. What did you see instead?
Configuration file stored at `/home/greg/.dlv`

### PR DESCRIPTION
**Pull request by GregorioMartinez**
This change checks for the environmental variable XDG_CONFIG_HOME
before storing the configuration file in the home directory. If found 
it stores the config at $XDG_CONFIG_HOME/dlv

Closes #1454

### CODE REVIEW

**Comment by GregorioMartinez** [comment]
You're right @aarzilli. 
I'll update this to default to `$HOME/.config` on Linux machines when `XDG_CONFIG_HOME` is not set.

For backwards compatibility what do you think would be best? Check for the existence of `~/.dlv` and move it to `XDG_CONFIG_HOME` if found? 

RE:  `XDG_CONFIG_DIRS` - This is used to search for more config files on top of `XDG_CONFIG_HOME`. I don't believe there would be a case where `/usr/local/share/:/usr/share/ ` or `/etc/xdg` would have an additional config for delve, but I can add it in if you think it's necessary.

> $XDG_DATA_DIRS defines the preference-ordered set of base directories to search for data files in addition to the $XDG_DATA_HOME base directory.

**Comment by aarzilli** [comment]
As I said before, as far as I'm concerned this is fine as it is.

**Comment by GregorioMartinez** [comment]
Alright, I can add the backwards compatibility flow in.
Should we also delete the current `$HOME/.dlv` directory or just print a message so users know they can remove it themselves?

**Comment by GregorioMartinez** [comment]
I made the requested changes, but not entirely sure how we should write the tests for this.
If anyone has suggestions on how to do this and could point me to the right direction that'd be helpful.

File: pkg/config/config.go [hunk 1]
```diff
@@ -68,6 +69,32 @@ func LoadConfig() *Config {
...
+		fmt.Printf("Successfully moved config location\n")
+		fmt.Printf("Attempting to remove old config directory: %s\n", oldLocation)
+
+		if err := os.RemoveAll(oldLocation); err != nil {
+			fmt.Printf("Unable to remove old config location: %s", err)
+			return &Config{}
+		}
+
+		fmt.Printf("Successfully removed old config directory\n")
+
```
Comment (aarzilli) [inline pkg/config/config.go]: Remove empty line.

File: pkg/config/config.go [hunk 2]
```diff
@@ -68,6 +69,32 @@ func LoadConfig() *Config {
...
 
+	h, err := hasOldConfig()
+	if err != nil {
+		fmt.Printf("Unable to determine if old config exists: %s", err)
+	}
+
+	if h {
+		userHomeDir := getUserHomeDir()
+		oldLocation := path.Join(userHomeDir, configDirHidden)
+		fmt.Printf("Attempting to move config from: %s/%s to %s\n", oldLocation, configFile, fullConfigFile)
```
Comment (aarzilli) [inline pkg/config/config.go]: I would prefer if this only print a maximum of one message (either one saying that the file was moved successfully or one saying that there was an error).

There's also the problem that this runs before the server starts listening which means that this messages being print before the "API server listening" message, which will break frontends.

Comment (GregorioMartinez) [reply pkg/config/config.go]: @aarzilli I consolidated the messages into one and removed a few other extra empty lines to better match the style that was already present in the file.

Can you elaborate on the API server listening message problem?
When I test it locally I get the message after the "API server listening at:" message and the current file also prints messages out in the same way that I am doing now.

Comment (aarzilli) [reply pkg/config/config.go]: The problem is that clients will look for the "API server listening at:" and expect it to be the first thing on stdout. I think those messages should go to stderr `fmt.Fprintf(os.Stderr, ...` so they don't interfere with that.

File: pkg/config/config.go [hunk 3]
```diff
@@ -71,19 +71,25 @@ func LoadConfig() *Config {
 
 	h, err := hasOldConfig()
 	if err != nil {
-		fmt.Printf("Unable to determine if old config exists: %s", err)
+		if _, err := fmt.Fprintf(os.Stderr, "Unable to determine if old config exists: %v\n", err); err != nil {
```
Comment (aarzilli) [inline pkg/config/config.go]: why? if stderr is closed, it's closed.

File: pkg/config/config.go [hunk 4]
```diff
@@ -71,25 +71,19 @@ func LoadConfig() *Config {
...
 		}
 
 		if err := os.RemoveAll(oldLocation); err != nil {
-			if _, err := fmt.Fprintf(os.Stderr, "Unable to remove old config location: %v\n", err); err != nil {
-				fmt.Printf("Unable to write to stderr: %v\n", err)
-			}
+			fmt.Fprintf(os.Stderr, "Unable to remove old config location: %v\n", err)
 			return &Config{}
 		}
 		fmt.Printf("Successfully moved config from: %s to: %s\n", oldLocation, fullConfigFile)
```
[the remark this answers is not in the document]
Comment (GregorioMartinez) [reply pkg/config/config.go]: @aarzilli Done. Thanks for catching that one.

**Comment by aarzilli** [comment]
LGTM

File: pkg/config/config.go [hunk 5]
```diff
@@ -67,6 +69,26 @@ func LoadConfig() *Config {
 		return &Config{}
 	}
 
+	h, err := hasOldConfig()
```
Comment (derekparker) [inline pkg/config/config.go]: nit: Let's use a more descriptive var name here than just `h`.

Comment (GregorioMartinez) [reply pkg/config/config.go]: @derekparker Done!

**Pull request merged**